eeps/eep-0078.md

```diff
@@ -0,0 +1,114 @@
+    Author: Michał Muskała <micmus(at)whatsapp(dot)com>
+    Status: Draft
+    Type: ...
+    Created: 16-12-2024
+    Erlang-Version: ...
+    Post-History: ...
+****
+EEP 78: Multi-valued comprehensions
+----
+
+Abstract
+========
+
+This EEP proposes enhancing the comprehension syntax to allow emitting multiple
+elements in a single iteration of the comprehension loop - effectively enhancing
+comprehensions to implement `flatmap` with a fixed number of elements.
+
+Rationale
+=========
+
+Comprehensions in Erlang are a very flexible and convinent way of implementing many
+iteration and looping patterns. However, there are some cases that end up unergonomic,
+notably, and where the scope of this EEP is, where we'd like to emit multiple elements
+from a single iteration.
+
+Existing forms of expressing this are awkward and introduce extra, unnecessary allocations.
+For example:
+
+    lists:append([[X + 1, X + 2] || X <- Xs]
+    [Tmp || X <- Xs, Tmp <- [X + 1, X + 2]]
+
+Both of those ways end up creating an extra allocation of a temporary 2-element list, introducing
+inefficiency, as well as are, arguably, harder to understand than necessary introducing extra function
+calls or variables.
+
+This EEP proposes enhancing comprehensions with the ability to do so using a very natural
+syntax extension of the existing comprehension syntax. In particular for list and map
+comprehensions:
+
+    [X + 1, X + 2, ... || X <- Xs]
+
+    #{K + 1 => V + 1, K + 2 => V + 2, ... || K := V <- Map}
+
+The semantics of map comprehensions where multiple keys in the same iteration would end up
+with the same value, should be the same as if the keys were emitted in subsequent iterations.
+
+Binary comprehensions already support this, and thus there's no enhancement to their syntax
+suggested in this EEP, for example:
+
+    << <<(X + 1), (X + 2)>> || <<X>> <= Bin>>.
```

Comment:
This is true, but also the multi-value syntax should be applicable to it, if only for the sake of consistency and completeness.

I think a binary comprehension like...
```erlang
<< <<(X + 1)>>, <<(X + 2)>> || <<X>> <= Bin >>.
```
... should be accepted and result in a concatenation of the generated binaries, that is, same as if written `<<(X + 1), (X + 2)>>`.

Reply:
Yes, that's a good point. TBH the syntax for binary comprehensions is quite noisy already and I didn't want to muddle it more